File: frontend/src/lib/sourcing.ts
/** Types + REST calls for sourcing (vendors, deals). */
import { api } from "./api";

export const VENDOR_TYPES = ["hotel", "transport", "activity"] as const;
export type VendorType = (typeof VENDOR_TYPES)[number];
export const DEAL_STATUSES = ["requested", "negotiating", "confirmed", "declined"] as const;
export type DealStatus = (typeof DEAL_STATUSES)[number];

export interface Vendor {
  id: string;
  name: string;
  type: VendorType;
  location: string | null;
  contact_email: string | null;
  contact_phone: string | null;
  notes: string | null;
}

export interface Deal {
  id: string;
  trip_id: string | null;
  vendor_id: string;
  kind: string;
  status: DealStatus;
  outreach_subject: string | null;
  outreach_body: string | null;
  sent: boolean;
  terms: string | null;
  amount: number | null;
  currency: string;
  created_at: string;
}

export const listVendors = () => api<Vendor[]>("/api/v1/vendors");
export const createVendor = (data: Partial<Vendor> & { name: string; type: VendorType }) =>
  api<Vendor>("/api/v1/vendors", { method: "POST", body: JSON.stringify(data) });
export const deleteVendor = (id: string) =>
  api<void>(`/api/v1/vendors/${id}`, { method: "DELETE" });
export const generateSourcing = (tripId: string) =>
  api<{ deals_created: number }>(`/api/v1/trips/${tripId}/sourcing/generate`, { method: "POST" });
export const listDeals = (tripId?: string) =>
  api<Deal[]>(`/api/v1/deals${tripId ? `?trip_id=${tripId}` : ""}`);
export const updateDeal = (id: string, data: { status?: DealStatus; terms?: string; amount?: number }) =>
  api<Deal>(`/api/v1/deals/${id}`, { method: "PATCH", body: JSON.stringify(data) });

File: backend/app/schemas/sourcing.py
"""Sourcing (vendors, deals) schemas."""
import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from app.models.sourcing import DEAL_STATUSES, VENDOR_TYPES


class VendorCreate(BaseModel):
    name: str = Field(min_length=1, max_length=160)
    type: str = Field(default="hotel", pattern="^(" + "|".join(VENDOR_TYPES) + ")$")
    location: str | None = None
    contact_email: str | None = None
    contact_phone: str | None = None
    notes: str | None = None


class VendorOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    name: str
    type: str
    location: str | None
    contact_email: str | None
    contact_phone: str | None
    notes: str | None


class DealOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    trip_id: uuid.UUID | None
    vendor_id: uuid.UUID
    kind: str
    status: str
    outreach_subject: str | None
    outreach_body: str | None
    sent: bool
    terms: str | None
    amount: int | None
    currency: str
    created_at: datetime


class DealUpdate(BaseModel):
    status: str | None = Field(default=None, pattern="^(" + "|".join(DEAL_STATUSES) + ")$")
    terms: str | None = None
    amount: int | None = None
